data/mongodb/v0/tasks.go

```diff
@@ -0,0 +1,437 @@
+	if err != nil {
+		return nil, err
+	}
+
+	outputStruct := DropDatabaseOutput{
+		Status: "Successfully dropped database",
+	}
+
+	output, err := base.ConvertToStructpb(outputStruct)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func (e *execution) createSearchIndex(ctx context.Context, in *structpb.Struct) (*structpb.Struct, error) {
+	var inputStruct CreateSearchIndexInput
+	err := base.ConvertFromStructpb(in, &inputStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	syntax := inputStruct.Syntax
+
+	searchIndexModel := mongo.SearchIndexModel{
+		Definition: syntax,
+		Options: &options.SearchIndexesOptions{
+			Name: &inputStruct.IndexName,
+			Type: &inputStruct.IndexType,
+		},
+	}
+
+	_, err = e.client.searchIndexClient.CreateOne(ctx, searchIndexModel)
+	if err != nil {
+		return nil, err
+	}
+
+	outputStruct := CreateSearchIndexOutput{
+		Status: "Successfully created search index",
+	}
+
+	// Convert the output structure to Structpb
+	output, err := base.ConvertToStructpb(outputStruct)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func (e *execution) dropSearchIndex(ctx context.Context, in *structpb.Struct) (*structpb.Struct, error) {
+	var inputStruct DropSearchIndexInput
+	err := base.ConvertFromStructpb(in, &inputStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	indexName := inputStruct.IndexName
+
+	err = e.client.searchIndexClient.DropOne(ctx, indexName)
+	if err != nil {
+		return nil, err
+	}
+
+	outputStruct := DropSearchIndexOutput{
+		Status: "Successfully dropped search index",
+	}
+
+	output, err := base.ConvertToStructpb(outputStruct)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func (e *execution) vectorSearch(ctx context.Context, in *structpb.Struct) (*structpb.Struct, error) {
+	var inputStruct VectorSearchInput
+	err := base.ConvertFromStructpb(in, &inputStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	exact := inputStruct.Exact
+	filter := inputStruct.Filter
+	indexName := inputStruct.IndexName
+	limit := inputStruct.Limit
+	numCandidates := inputStruct.NumCandidates
+	path := inputStruct.Path
+	queryVector := inputStruct.QueryVector
+	fields := inputStruct.Fields
+
+	vectorSearch := bson.M{
+		"exact":       exact,
+		"index":       indexName,
+		"path":        path,
+		"queryVector": queryVector,
+		"limit":       limit,
+	}
+	if filter != nil {
+		vectorSearch["filter"] = filter
+	}
+
+	if !exact {
+		if numCandidates > 0 {
+			vectorSearch["numCandidates"] = numCandidates
+		} else {
+			vectorSearch["numCandidates"] = 3 * limit
+		}
+	}
+
+	project := bson.M{"_id": 0}
+	for _, field := range fields {
+		project[field] = 1
+	}
+
+	query := bson.A{
+		bson.M{
+			"$vectorSearch": vectorSearch,
+		},
+		bson.M{
+			"$addFields": bson.M{
+				"score": bson.M{
+					"$meta": "vectorSearchScore",
+				},
+			},
+		},
+	}
+
+	if len(fields) > 0 {
+		query = append(query, bson.M{
+			"$project": project,
+		})
+	}
```

Comment:
Important
About the logic above, I am not so sure the logic about this part.
```
	if !exact {
		if numCandidates > 0 {
			vectorSearch["numCandidates"] = numCandidates
		} else {
			vectorSearch["numCandidates"] = 3 * limit
		}
	}

	project := bson.M{"_id": 0}
	for _, field := range fields {
		project[field] = 1
	}

```

1. when exact is true, what should it be for numCandidates? 
2. why do we need to set `project[field] = 1`? What does it mean here?

I am not so familiar with this part, please make sure the function works as expectation!

Reply:
@chuang8511 

1. According to mongodb atlas vector search docs,`exact :false - to run ANN search, true - to run ENN search`. Basically ANN doesn't consider every data and uses some efficient algorithm instead, which is pretty good for larger scale collection (> 10000 documents is preferable according to mongodb, that's why numCandidates is needed. Whilst ENN search consider every data within a collection (<= 10000 documents is preferable), numCandidates isnt needed as every documents is being considered within the vector search.

2. In mongodb aggregate, if we want to filter the retrieved documents with specific fields, we have to construct map that looks like
`"$project":{
"field1" : 1,
"field2" : 1
}`
where 1 means true and 0 means false

Reply:
@chuang8511 

And one more thing is that I put _id : 0 in order not to show id field by default, since we talk about data migration, sometimes the user doesn't really need such metadata, but still if the user insist on including _id, they still can do it via fields input, for instance ["name", "age", "_id"]